### PR TITLE
Support wasmJs

### PR DESCRIPTION
**Pull request by ahna92**
**Changes** 

* Add wasmJs target
* Update kotlinx-atomicfu plugin 
* Update kotlinx-serialization-json
* Update kotlinx-coroutines

### CODE REVIEW

**Comment by snappdevelopment** [comment]
Hi, thanks for your PR!
I will review your changes in the upcoming days. I'll need to read about how wasm/js works, because I am not a web dev and don't really know how that target works. I might come back with some questions as well.

**Comment by ahna92** [comment]
@snappdevelopment yes ofc take your time 

I've added a run config for webApp too 

the library it self wont be affected by the wasm changes (just adding the new target)

most of the web changes i made were on the sample project 

official [example](https://github.com/JetBrains/compose-multiplatform/tree/master/examples/imageviewer)

**Comment by ahna92** [comment]
@snappdevelopment Also would you conceder entering [hacktoberfest](https://hacktoberfest.com/) (just need to add `hacktoberfest` to project topics)
This is something random has nothing to do with the PR :P

File: gradle.properties [hunk 1]
```diff
@@ -42,4 +46,9 @@ POM_SCM_DEV_CONNECTION=scm:git:ssh://git@github.com/snappdevelopment/JsonTree.gi
 
 POM_DEVELOPER_ID=snappdevelopment
 POM_DEVELOPER_NAME=Sebastian Neubauer
-POM_DEVELOPER_URL=https://github.com/snappdevelopment/
\ No newline at end of file
+POM_DEVELOPER_URL=https://github.com/snappdevelopment/
+
+# wasm falgs
+org.jetbrains.compose.experimental.jscanvas.enabled=true
+org.jetbrains.compose.experimental.macos.enabled=true
```
Comment (snappdevelopment) [inline gradle.properties:53]: Do we need this flag? From a Google search it seems like this is for an experimental native macOS target. I haven't tried that yet. I don't think we need it for now.

File: sample/build.gradle.kts [hunk 2]
```diff
@@ -54,6 +75,16 @@ kotlin {
...
+
+        val jsWasmMain by creating {
+            dependsOn(commonMain)
+        }
+        val jsMain by getting {
+            dependsOn(jsWasmMain)
+        }
+        val wasmJsMain by getting {
+            dependsOn(jsWasmMain)
+        }
```
Comment (snappdevelopment) [inline sample/build.gradle.kts:87]: I don't we need to define those source sets. I tried to run the app without it and it worked. I think every target depends automatically on commonMain.
I downloaded the current template from https://kmp.jetbrains.com/. It also doesn't define those source sets.

File: sample/build.gradle.kts [hunk 3]
```diff
@@ -26,20 +27,40 @@ kotlin {
...
         }
-        commonMain.dependencies {
-            implementation(compose.runtime)
-            implementation(compose.foundation)
-            implementation(compose.material3)
-            implementation(compose.components.resources)
-            implementation(compose.components.uiToolingPreview)
-            implementation(libs.kotlinx.serialization.json)
-            implementation(project(":jsontree"))
+        val commonMain by getting {
```
Comment (snappdevelopment) [inline sample/build.gradle.kts:54]: Considering my comment below, we can revert this change as well

File: settings.gradle.kts [hunk 4]
```diff
@@ -6,7 +6,7 @@ pluginManagement {
     }
 }
 dependencyResolutionManagement {
-    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
+//    repositoriesMode.set(RepositoriesMode.FAIL_ON_PROJECT_REPOS)
```
Comment (snappdevelopment) [inline settings.gradle.kts:9]: This can be reverted, we still want that line

File: .idea/runConfigurations/webApp.xml [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<component name="ProjectRunConfigurationManager">
```
Comment (snappdevelopment) [inline .idea/runConfigurations/webApp.xml:1]: Can we also add a `jsApp` run config? It should run `:sample:jsBrowserRun`. 
This one can be renamed to `wasmJsApp`.

**Review by snappdevelopment — CHANGES_REQUESTED** [review]

File: sample/build.gradle.kts (hunk 2 above)
Comment (ahna92) [reply sample/build.gradle.kts:87]: Sometime `wasmJs` and `Js` shared code (javascript browser code) so we create this common js target 

the sample does not have a common code right now so it will pass 

but you can see this usage in this [jetbrains example](https://github.com/JetBrains/compose-multiplatform/blob/1b2669f59d8f9a5cda1a26388a51e8c6793981c8/examples/imageviewer/shared/build.gradle.kts#L64)

I would recommend keeping it, but i can remove no issue up to you

File: settings.gradle.kts (hunk 4 above)
Comment (ahna92) [reply settings.gradle.kts:9]: `js` target imports libraries dynamically from the plugin it self  (npm packages)

So build was failing with this line, 
I tried using preferred settings , also failed so i commented

File: gradle.properties (hunk 1 above)
Comment (ahna92) [reply gradle.properties:53]: I think can remove np

File: settings.gradle.kts (hunk 4 above)
Comment (snappdevelopment) [reply settings.gradle.kts:9]: True, I also get an error now. Then we can delete this line.

File: sample/build.gradle.kts (hunk 2 above)
Comment (snappdevelopment) [reply sample/build.gradle.kts:87]: I see. That makes sense, let's keep it.

File: gradle.properties (hunk 1 above)
Comment (snappdevelopment) [reply gradle.properties]: Let's remove it for now

File: .idea/runConfigurations/webApp.xml (hunk 5 above)
Comment (snappdevelopment) [reply .idea/runConfigurations/webApp.xml]: I think you forgot to add the jsApp config

Comment (snappdevelopment) [reply .idea/runConfigurations/webApp.xml]: 🤔 If that's the case, then I am confused. From what I read, I thought that Kotlin/Js and Kotlin/wasm are two different targets (both web targets though). Kotlin/Js instructions are [here](https://kotlinlang.org/docs/js-project-setup.html) and Kotlin/wasm instructions [here](https://kotlinlang.org/docs/wasm-get-started.html). Kotlin/Js is considered stable and has it's own task `./gradlew jsRun`, while Kotlin/wasm is still experimental and runs with `./gradlew wasmJsBrowserRun`. Or am I missing something?

Comment (ahna92) [reply .idea/runConfigurations/webApp.xml]: `Kotlin/Js` uses html to build UI (stable)
`Kotlin/Wasm` uses compose to build UI (alpha)

business logic can be shared between the two 

the most inclusive setup is to support both of them and use kotlin/wasm for UI 

can review this [example](https://github.com/JetBrains/compose-multiplatform/tree/master/examples/imageviewer#run-on-web-via-gradle)

Comment (snappdevelopment) [reply .idea/runConfigurations/webApp.xml]: I see. So when someone builds a Kotlin/Js project and uses this library (or when running `./gradlew jsRun` on the sample) it will still run the wasm app. That's why `main.js.kt` uses `onWasmReady {}`.

Comment (ahna92) [reply .idea/runConfigurations/webApp.xml]: yes

**Comment by snappdevelopment** [comment]
Looks good! Thanks for the contribution, I will merge this now.

**Review by snappdevelopment — APPROVED** [review]
